src/main/java/com/amazonaws/encryptionsdk/kms/AwsKmsDataKeyEncryptionDaoBuilder.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except
+ * in compliance with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package com.amazonaws.encryptionsdk.kms;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.services.kms.AWSKMSClientBuilder;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+
+/**
+ * Builder to construct an AwsKmsDataKeyEncryptionDao.
+ * CredentialProvider and ClientConfiguration are optional
+ * and may be configured if necessary.
+ */
+public class AwsKmsDataKeyEncryptionDaoBuilder {
+
+    private AWSKMSClientBuilder awsKmsClientBuilder;
+    private AWSCredentialsProvider credentialsProvider;
+    private ClientConfiguration clientConfiguration;
+    private List<String> grantTokens;
+    private String regionId;
+
+    // The user agent string is used to note the AWS Encryption SDK's language and version in calls to AWS KMS
+    // Since the AWS KMS client is being constructed the AWS Encryption SDK, we can append this value
+    // unless a custom client configuration was provided
+    private boolean canAppendUserAgentString = true;
+
+    /**
+     * A builder to construct the default AwsKmsDataKeyEncryptionDaoBuilder that will create clients
+     * for an AWS region. Credentials, client configuration, and grant tokens may be specified if necessary.
+     *
+     * @return The AwsKmsDataKeyEncryptionDaoBuilder
+     */
+    public static AwsKmsDataKeyEncryptionDaoBuilder defaultBuilder() {
+        return new AwsKmsDataKeyEncryptionDaoBuilder(AWSKMSClientBuilder.standard());
+    }
+
+    AwsKmsDataKeyEncryptionDaoBuilder(AWSKMSClientBuilder awsKmsClientBuilder) {
+        this.awsKmsClientBuilder = awsKmsClientBuilder;
+    }
+
+    public AwsKmsDataKeyEncryptionDao build() {
+        if (credentialsProvider != null) {
+            awsKmsClientBuilder = awsKmsClientBuilder.withCredentials(credentialsProvider);
+        }
+
+        if (clientConfiguration != null) {
+            canAppendUserAgentString = false;
```

Comment:
Do we want this behavior? Previously I thought we would still pass the user agent string even if the user specified a custom client configuration. i.e. we're still creating the kms client here, even if the user is specifying the configuration.

Reply:
I believe so. From offline discussions, it is safer to not append the user agent string if a client configuration is provided. I don't believe we can really tell if a customer has a specific user agent string they want to use otherwise (or if they manually cleared it themselves).